README.md
# FILE_BASED_CMS_170

Content Management System Project for Course 170 in Launch School

Features:
- CRUD application that uses Ruby to interact with file system
- A directory with tests for each feature in the application
- Uses YML file to log users and passwords using bcrypt encryption
- Uses puma as local webserver
- Uses Sinatra web framework for routes and layout
- Rakefile for easy access of running tests

## Installation and Use
1. Download the zip files and unzip into a new directory.
2. To run the application, open Terminal,  ```cd``` into the directory, and type ```ruby cms.rb```.
3. Leave the server running and open a web browser.
4. Navigate to ```localhost:4567``` to interact with the application.

HINT: In order to log in, use ```admin``` as your username and ```secret``` as your password.


data/changes.txt
I saved this in the web app!
I just deleted a bunch of other stuff and then wrote this.

Some additions.

test/cms_test.rb
ENV["RACK_ENV"] = "test"

require "minitest/autorun"
require "rack/test"
require "fileutils"
require "yaml"

require_relative "../cms"

class AppTest < Minitest::Test
  include Rack::Test::Methods

  def app
    Sinatra::Application
  end

  def setup
    FileUtils.mkdir_p(data_path)
  end

  def teardown
    FileUtils.rm_rf(data_path)
  end

  def session
    last_request.env["rack.session"]
  end

  def create_document(name, content = "")
    File.open(File.join(data_path, name), "w") do |file|
      file.write(content)
    end
  end

  def admin_session
    { "rack.session" => { username: "admin" } }
  end

  def test_index
    create_document "about.md"
    create_document "changes.txt"
    create_document "something_different.txt"


    get "/"

    assert_equal 200, last_response.status
    assert_equal "text/html;charset=utf-8", last_response["Content-Type"]
    assert_includes last_response.body, "about.md"
    assert_includes last_response.body, "changes.txt"
    assert_includes last_response.body, "something_different.txt"
  end

  def test_viewing_text_document
    create_document "changes.txt", "Changing"

    get "/changes.txt"

    assert_equal 200, last_response.status
    assert_equal "text/plain", last_response["Content-Type"]
    assert_includes last_response.body, "Changing"
  end

  def test_viewing_markdown_document
    create_document "about.md", '# Ruby is...'
    get "/about.md"

    assert_equal 200, last_response.status
    assert_equal "text/html;charset=utf-8", last_response["Content-Type"]
    assert_includes last_response.body, "<h1>Ruby is...</h1>"
  end

  def test_error_message_for_invalid_file
    create_document "isafile.txt"

    get "/notafile.txt"
    assert_equal 302, last_response.status
    assert_equal "notafile.txt does not exist! Try one of these files:", session[:message]
  end

  def test_editing_document_signed_in
    create_document "changes.txt", "changes"
    get "/changes.txt/edit", {}, admin_session

    assert_equal 200, last_response.status
    assert_includes last_response.body, "<textarea"
    assert_includes last_response.body, '<input type="submit"'
  end

  def test_editing_document_signed_out
    create_document "changes.txt", "changes"
    get "/changes.txt/edit"

    assert_equal 302, last_response.status
    assert_equal "You must be signed in to do that.", session[:message]
  end

  def test_updating_document_signed_in
    create_document "changes.txt"
    post "/changes.txt/update", { content: "Changing" }, admin_session

    assert_equal 302, last_response.status
    assert_equal "changes.txt has been updated successfully!", session[:message]

    get "/changes.txt"
    assert_equal 200, last_response.status
    assert_includes last_response.body, "Changing"
  end

  def test_updating_document_signed_out
    create_document "changes.txt"
    post "/changes.txt/update", content: "Changing"

    assert_equal 302, last_response.status
    assert_equal "You must be signed in to do that.", session[:message]
  end

  def test_new_document_button_on_index_page
    get "/"
    assert_equal 200, last_response.status
    assert_includes last_response.body, "New Document</a>"
  end

  def test_view_new_document_form_signed_in
    get "/new_file/create", {}, admin_session

    assert_equal 200, last_response.status
    assert_includes last_response.body, "<input"
    assert_includes last_response.body, %q(<input type="submit")
  end

  def test_view_new_document_form_signed_out
    get "/new_file/create"

    assert_equal 302, last_response.status
    assert_equal "You must be signed in to do that.", session[:message]
  end

  def test_create_new_document_signed_in
    post "/new_file/save", { file: "test.txt" }, admin_session
    assert_equal 302, last_response.status
    assert_equal "test.txt was created.", session[:message]

    get "/"
    assert_includes last_response.body, "test.txt"
  end

  def test_create_new_document_signed_out
    post "/new_file/save", file: "test.txt"
    assert_equal 302, last_response.status
    assert_equal "You must be signed in to do that.", session[:message]
  end

  def test_create_new_document_without_filename
    post "/new_file/save", { file: "" }, admin_session
    assert_equal 422, last_response.status
    assert_includes last_response.body, "Invalid filename"
  end

  def test_deleting_document_signed_in
    create_document("test.txt")

    post "/test.txt/delete", {}, admin_session

    assert_equal 302, last_response.status
    assert_equal "test.txt was deleted.", session[:message]

    get "/"
    refute_includes last_response.body, "href=\"test.txt\""
  end

  def test_deleting_document_signed_out
    create_document("test.txt")

    post "/test.txt/delete"

    assert_equal 302, last_response.status
    assert_equal "You must be signed in to do that.", session[:message]
  end

  def test_signin_form
    get "/users/signin"

    assert_equal 200, last_response.status
    assert_includes last_response.body, "<input"
    assert_includes last_response.body, %q(<input type="submit")
  end

  def test_signin
    post "/users/authenticate", username: "admin", password: "secret"
    assert_equal 302, last_response.status
    assert_equal "Welcome!", session[:message]
    assert_equal "admin", session[:username]

    get last_response["Location"]
    assert_includes last_response.body, "Signed in as admin"
  end

  def test_signin_with_bad_credentials
    post "/users/authenticate", username: "guest", password: "shhhh"
    assert_equal 422, last_response.status
    assert_nil session[:username]
    assert_includes last_response.body, "Invalid credentials"
  end

  def test_signout
    get "/", {}, {"rack.session" => { username: "admin" } }
    assert_includes last_response.body, "Signed in as admin"

    post "/users/signout"
    get last_response["Location"]

    assert_nil session[:username]
    assert_includes last_response.body, "You have been signed out"
    assert_includes last_response.body, "Sign In"
  end
end


cms.rb
require "sinatra"
require "sinatra/reloader"
require "tilt/erubis"
require "bcrypt"
require "redcarpet"
require "fileutils"
require "yaml"
require 'pry'

configure do
  enable :sessions
  set :session_secret, 'secret'
end


def data_path
  if ENV["RACK_ENV"] == "test"
    Dir.pwd + "/test/data"
  else
    Dir.pwd + "/data"
  end
end

def data_filenames
  Dir.new(data_path).select { |file| file.match?(/.+\..+/) }
end

def valid_file?(filename)
  data_filenames.include? filename
end

def valid_filename?(filename)
  filename.match?(/.+\..+/)
end

def render_markdown(file)
  markdown = Redcarpet::Markdown.new(Redcarpet::Render::HTML)
  markdown.render(file.read)
end

def parse_content(path, pure_content=nil)
  extension = path.split(".").last
  extension = "txt" if pure_content == :pure_content
  file = File.new(path)
  case extension
  when "md"
    render_markdown(file)
  when "txt"
    headers["Content-Type"] = "text/plain"
    file.read
  end
end

def load_users_and_passwords
  path = if ENV["RACK_ENV"] == "test"
    File.expand_path("../test/users.yml", __FILE__)
  else
    File.expand_path("../users.yml", __FILE__)
  end
  YAML.load_file(path)
end

def valid_username_and_password?(username, password)
  credentials = load_users_and_passwords
  
  if credentials.key?(username)
    bcrypt_password = BCrypt::Password.new(credentials[username])
    bcrypt_password == password
  else
    false
  end
end

def signed_in?
  !!session[:username]
end

def redirect_user_if_not_signed_in
  if !signed_in?
    session[:message] = "You must be signed in to do that."
    redirect "/"
  end
end

# List all files in the data directory
get "/" do
  @filenames = data_filenames

  erb :index, layout: :layout
end

# Show the contents of a specific file
get "/:filename" do
  @filename = params[:filename]
  if valid_file?(@filename)
    path = File.join(data_path, @filename)
    parse_content(path)
  else
    session[:message] = "#{@filename} does not exist! Try one of these files:"
    redirect "/"
  end
end

# Display form to edit contents of file
get "/:filename/edit" do
  redirect_user_if_not_signed_in
  @filename = params[:filename]
  if valid_file?(@filename)
    @path = File.join(data_path, @filename)
    @content = parse_content(@path, :pure_content)
    headers["Content-Type"] = "text/html"
    erb :edit
  else
    session[:message] = "#{@filename} does not exist! Try one of these files:"
    redirect "/"
  end
end

# Update/save changes made to file
post "/:filename/update" do
  redirect_user_if_not_signed_in
  filename = params[:filename]
  content = params[:content]
  path = File.join(data_path, filename)
  File.write(path, content)
  session[:message] = "#{filename} has been updated successfully!"
  redirect "/"
end

# Create a new file
get "/new_file/create" do
  redirect_user_if_not_signed_in
  erb :create, layout: :layout
end

# Save a valid new file
post "/new_file/save" do
  redirect_user_if_not_signed_in
  if params[:file]
    filename = params[:file]
    if valid_filename?(filename)
      File.open(File.join(data_path, filename), "w") {}
      session[:message] = "#{filename} was created."
      status 302
      redirect "/"
    else
      session[:message] = "Invalid filename!"
      status 422
      erb :create, layout: :layout
    end
  end
end

post "/:filename/delete" do
  redirect_user_if_not_signed_in
  filename = params[:filename]
  if valid_filename?(filename)
    File.delete(File.join(data_path, filename))
    session[:message] = "#{filename} was deleted."
    redirect "/"
  else
    session[:message] = "#{filename} does not exist! Try one of these files:"
    redirect "/"
  end
end

get "/users/signin" do
  erb :signin, layout: :layout
end

post "/users/authenticate" do
  if valid_username_and_password?(params[:username], params[:password])
    session[:username] = params[:username]
    session[:message] = "Welcome!"
    redirect "/"
  else
    session[:message] = "Invalid credentials"
    status 422
    erb :signin, layout: :layout
  end
end

post "/users/signout" do
  session[:message] = "You have been signed out."
  session.delete(:username)
  redirect "/"
end
